chairman/Chairman.hs

```diff
@@ -0,0 +1,434 @@
+       )
+    => CoreNodeId
+    -> ChainsVar m blk
+    -> blk
+    -> STM m ()
+addBlock coreNodeId chainsVar blk =
+    modifyTVar' chainsVar (Map.adjust (CF.addBlock blk) coreNodeId)
+
+
+data ChairmanError blk = NodeMisconduct [blk]
+
+instance Condense blk => Show (ChairmanError blk) where
+    show (NodeMisconduct blks) = "NodeMisconduct " ++ condense blks
+
+instance (Condense blk, Typeable blk) => Exception (ChairmanError blk)
+
+
+-- | Check if the oldest immutable tip agree, keep forks from the last common
+-- block (including it).
+--
+checkAndSubtract
+    :: forall blk m.
+       ( MonadSTM m
+       , HasHeader blk
+       , Condense blk
+       )
+    => ChainsVar m blk
+    -> Int
+    -> STM m (ChairmanTrace blk)
+checkAndSubtract chainsVar securityParam = do
+    chains <- readTVar chainsVar
+    let (res, chains') = subtractPure chains
+    writeTVar chainsVar $! chains'
+    return res
+  where
+    -- it isn't really pure as it throws the 'NodeMisconduct' exception
+    subtractPure :: Map CoreNodeId (ChainFragment blk)
+                 -> ( ChairmanTrace blk
+                    , Map CoreNodeId (ChainFragment blk)
+                    )
+    subtractPure chains =
+      let tips :: [Maybe blk]
+          tips = foldl' (\tips' cf -> CF.head cf : tips') [] chains
+
+          -- find intersection of all the chains; the CLI guarantees that
+          -- there's at least one entry in @chains@, thus @fold1@ is safe here.
+          common :: ChainFragment blk
+          common =
+            foldl1
+              (\common' cf ->
+                case CF.intersectChainFragments common' cf of
+                  Nothing -> CF.Empty
+                  Just (_, common'', _, _)
+                          -> common'')
+              chains
+
+          mbHeadSlot = CF.headSlot common
+
+          -- remove common chain fragment from all the chains, but preserve the
+          -- common intersection point, this way we avoid leaving empty chains.
+          !chains'  = case mbHeadSlot of
+            Nothing       -> chains
+            Just headSlot ->
+                  (CF.takeWhileNewest (\blk -> Block.blockSlot blk >= headSlot))
+              <$> chains
+
+      in if minimum (CF.length <$> chains') <= securityParam
+            then case mbHeadSlot of
+                    -- but there is no intersection, but forks are short;  This
+                    -- might happen when starting and the nodes have not yet
+                    -- found consensus.
+                    Nothing ->
+                      ( NotFoundCommonBlock tips
+                      , chains'
+                      )
+                    -- there is an intersection and all forks are short
+                    Just headSlot ->
+                      ( WitnessedConsensusAt headSlot (map Block.blockPoint $ catMaybes tips)
+                      , chains'
+                      )
+            -- there is a long fork
+            else throw $ NodeMisconduct tips
```

Comment:
What monad is this throwing in?

Reply:
here it throws in pure code, it's wrapped in `MonadSTM m`, which at runtime will be `STM`.

Reply:
There is a comment `it isn't really pure as it throws the 'NodeMisconduct' exception`. I would advise using `Except` or `Either` for this function and then throwing the exception in `STM` (use `MonadThrow (STM m`).

Reply:
Done.